azimuth/modules/word_analysis/top_words.py

```diff
@@ -102,49 +101,57 @@ def compute_on_dataset_split(self) -> List[TopWordsResponse]:  # type: ignore
+                        for word, _ in filter(
+                            lambda s: s[1] > importance_saliency, zip(words, saliencies)
+                        )
+                    ]
+        # If saliency is not available, we proxy important words as any word that is neither
+        # punctuation or a stop word.
+        else:
+            spacy_model = spacy.load(self.config.syntax.spacy_model)
+            utterances = ds[self.config.columns.text_input]
+            for idx, utterance in enumerate(utterances):
+                doc = spacy_model(clean_utterance(utterance))
+                important_words_per_idx[idx] = [
+                    token.text for token in doc if not token.is_stop and not token.is_punct
+                ]
 
-        tokenizer = self.artifact_manager.get_tokenizer()
         is_error = np.array(
             get_predictions_from_ds(ds, self.mod_options.without_postprocessing)
         ) != np.array(ds[self.config.columns.label])
 
-        for idx, record in enumerate(words_saliencies):
-            # Put everything to lower case and remove cls/sep tokens.
-            words, saliencies = zip(
-                *[
-                    (word.lower(), saliency_value)
-                    for word, saliency_value in zip(record.words, record.saliency)
-                    if word not in [tokenizer.cls_token, tokenizer.sep_token]
-                ]
-            )
-            if importance_criteria == TopWordsImportanceCriteria.salient and words != []:
-                th_importance = self.mod_options.th_importance
-                assert len(words) == len(saliencies)
-                importance_saliency = max(th_importance * max(record.saliency), MIN_SALIENCY)
-                important_words = [
-                    word
-                    for word, _ in filter(
-                        lambda s: s[1] > importance_saliency, zip(words, saliencies)
-                    )
-                ]
-            # If saliency is not available, we proxy important words as any word that is neither
-            # punctuation or a stop word.
-            else:
-                important_words = [
-                    word for word in words if word not in self.stop_words_punctuation
-                ]
-
-            important_words_all.extend(important_words)
-            if is_error[idx]:
-                important_words_errors.extend(important_words)
-            else:
-                important_words_right.extend(important_words)
+        important_words_all = [word for words in important_words_per_idx.values() for word in words]
+        important_words_right = [
+            word
+            for idx, words in important_words_per_idx.items()
+            for word in words
+            if not is_error[idx]
+        ]
+        important_words_errors = [
+            word
+            for idx, words in important_words_per_idx.items()
+            for word in words
+            if is_error[idx]
+        ]
```

Comment:
In this case, I didn't hate the single loop that populates the 3 lists. It seems less repetitive. That said, I don't have a string preference. What do you think?
```suggestion
        important_words_all = []
        important_words_right = []
        important_words_errors = []
        for idx, words in important_words_per_idx.items():
            important_words_all.extend(words)
            if is_error[idx]:
                important_words_errors.extend(words)
            else:
                important_words_right.extend(words)
```

Reply:
Yeah, you are right. list comprehensions are not always the best when you are looping over the same items.